controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,362 @@
+			Termination: routev1.TLSTerminationEdge,
+		}
+	}
+
+	// Check if the certificate is up-to-date and not expired.
+	upToDate, renewalTime := func(route *routev1.Route) (bool, time.Time) {
+		pk, err := cmutilpki.DecodePrivateKeyBytes([]byte(route.Spec.TLS.Key))
+		if err != nil {
+			log.V(5).Info("Route has missing or invalid private key", "err", err)
+			return false, time.Time{}
+		}
+		certs, err := cmutilpki.DecodeX509CertificateChainBytes([]byte(route.Spec.TLS.Certificate))
+		if err != nil {
+			log.V(5).Info("Route has missing or invalid certificate", "err", err)
+			return false, time.Time{}
+		}
+		// if no error, certs[0] is the leaf cert (or self-signed cert)
+		matches, err := cmutilpki.PublicKeyMatchesCertificate(pk, certs[0])
+		if err != nil {
+			log.V(5).Info("Couldn't check if Route private key matches Route certificate", "err", err)
+			return false, time.Time{}
+		}
+		if !matches {
+			log.V(5).Info("Route private key doesn't match route cert")
+			return false, time.Time{}
+		}
+		// if we get here, we have a matching TLS cert/key pair.
+		// See if the cert will expire soon.
+		start := certs[0].NotBefore
+		end := certs[0].NotAfter
+		lifetime := end.Sub(start)
+		remaining := end.Sub(time.Now())
+		if remaining <= lifetime/3 {
+			return false, time.Time{}
+		}
+		return true, start.Add(2 * lifetime / 3)
+	}(instance)
+
+	if upToDate {
+		return c.ManageSuccessWithRequeue(ctx, instance, renewalTime.Sub(time.Now()))
+	}
+
+	// If we get here we need to fire off a certificate request. First step is to generate a new
+	// private key.
+	nextPrivateKeyPEM, found := instance.GetAnnotations()[cmapi.IsNextPrivateKeySecretLabelKey]
+	if !found {
+		// Generate a new private key, then re-reconcile.
+		// TODO: make key type and size configurable with an annotation
+		pk, err := cmutilpki.GenerateRSAPrivateKey(cmutilpki.MinRSAKeySize)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		pkPEM, err := cmutilpki.EncodePrivateKey(pk, cmapi.PKCS1)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		instance.GetAnnotations()[cmapi.IsNextPrivateKeySecretLabelKey] = string(pkPEM)
+		err = c.GetClient().Update(ctx, instance)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
+	}
+	// Is the Private Key malformed?
+	pk, err := cmutilpki.DecodePrivateKeyBytes([]byte(nextPrivateKeyPEM))
+	if err != nil {
+		log.V(5).Error(err, "next private key annotation doesn't contain a valid key")
+		delete(instance.GetAnnotations(), cmapi.IsNextPrivateKeySecretLabelKey)
+		err = c.GetClient().Update(ctx, instance)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
+	}
+
+	// We have a private key. Look for matching CertificateRequests
+	certificateRequests := &cmapi.CertificateRequestList{}
+	err = c.GetClient().List(ctx, certificateRequests, client.InNamespace(instance.Namespace))
```

Comment:
Does it not make sense to label this route's CertificateRequest's with a label containing the route's name? This would reduce the search space here.